src/embeddedTraceViewer.ts

```diff
@@ -0,0 +1,260 @@
+import { getNonce } from './utils';
+import * as vscodeTypes from './vscodeTypes';
+import { DisposableBase } from './disposableBase';
+import { PlaywrightTestServer } from './playwrightTestServer';
+import { TraceViewer } from './traceViewer';
+import { SettingsModel } from './settingsModel';
+import { PlaywrightTestCLI } from './playwrightTestCLI';
+
+export class EmbeddedTraceViewer implements TraceViewer {
+  readonly vscode: vscodeTypes.VSCode;
+  readonly extensionUri: vscodeTypes.Uri;
+  private _settingsModel: SettingsModel;
+  private _currentFile?: string;
+  private _traceViewerPanel: EmbeddedTraceViewerPanel | undefined;
+  private _traceLoadRequestedTimeout?: NodeJS.Timeout;
+  private _config: TestConfig;
+  private _testServer?: PlaywrightTestServer;
+
+  constructor(vscode: vscodeTypes.VSCode, settingsModel: SettingsModel, extensionUri: vscodeTypes.Uri, config: TestConfig, testServer?: PlaywrightTestServer | PlaywrightTestCLI) {
+    this.vscode = vscode;
+    this.extensionUri = extensionUri;
+    this._settingsModel = settingsModel;
+
+    this._config = config;
+    if (testServer instanceof PlaywrightTestServer)
+      this._testServer = testServer;
+  }
+
+  isStarted() {
+    return !!this._traceViewerPanel;
+  }
+
+  currentFile() {
+    return this._currentFile;
+  }
+
+  async willRunTests() {
+    await this._startIfNeeded();
+  }
+
+  async open(file: string) {
+    await this._startIfNeeded();
+    this._currentFile = file;
+    this._fireLoadTraceRequestedIfNeeded();
+  }
+
+  close() {
+    this._traceViewerPanel?.dispose();
+    this._traceViewerPanel = undefined;
+    this._currentFile = undefined;
+    if (this._traceLoadRequestedTimeout) {
+      clearTimeout(this._traceLoadRequestedTimeout);
+      this._traceLoadRequestedTimeout = undefined;
+    }
+  }
+
+  private _fireLoadTraceRequestedIfNeeded() {
+    if (this._traceLoadRequestedTimeout) {
+      clearTimeout(this._traceLoadRequestedTimeout);
+      this._traceLoadRequestedTimeout = undefined;
+    }
+    if (!this._traceViewerPanel || !this._currentFile)
+      return;
+    this._traceViewerPanel.postMessage({ method: 'loadTraceRequested', params: { traceUrl: this._currentFile } });
+    if (this._currentFile.endsWith('.json'))
+      this._traceLoadRequestedTimeout = setTimeout(() => this._fireLoadTraceRequestedIfNeeded(), 500);
+  }
+
+  private async _computeServerUrlPrefix() {
+    if (!this._testServer)
+      return;
+    const wsEndpoint = await this._testServer.wsEndpoint();
+    if (!wsEndpoint)
+      return;
+    const serverUrl = new URL(wsEndpoint);
+    serverUrl.protocol = 'http';
+    if (this.vscode.env.remoteName && ['[::1]', '0.0.0.0'].includes(serverUrl.hostname))
+      serverUrl.hostname = 'localhost';
+    const serverUri = await this.vscode.env.asExternalUri(this.vscode.Uri.parse(serverUrl.origin));
+    return serverUri.toString().replace(/\/$/, '');
+  }
+
+  private async _startIfNeeded() {
+    if (this._traceViewerPanel)
```

Comment:
This code has inherent race condition where calling it 5 times will create 5 panel instances.

Reply:
Addressed in 7b384dbe30b8ccc1fde6d310adb668c8a9421e65